client/gutenberg/editor/controller.js

```diff
@@ -43,13 +55,20 @@ export const post = ( context, next ) => {
 	const unsubscribe = context.store.subscribe( () => {
 		const state = context.store.getState();
 		const siteId = getSelectedSiteId( state );
+		const userId = getCurrentUserId( state );
 
 		if ( ! siteId ) {
 			return;
 		}
 
 		unsubscribe();
 
+		if ( ! userId ) {
+			return;
+		}
+
+		registerDataPlugins( userId );
```

Comment:
The trouble is that this plugin registration call comes too late, at a time when the `@wordpress/core-data` store is already created and registered. Without the appropriate middlewares.

The `@wordpress/core-data` store is created during the static initialization of the module, i.e., when the import statement
```js
import '@wordpress/core-data';
```
is evaluated.

That happens earlier in this file when importing the `gutenberg/editor/main` module. That module imports `@wordpress/core-data` and even if it didn't, importing anything from `@wordpress/editor` will recursively import (an initialize) `core-data`.

The only way to escape the static initialization is to import the `gutenberg/editor/main` using a `require()` call after calling `registerDataPlugins`. That's what I did in https://github.com/Automattic/wp-calypso/pull/27778/commits/bf78286958e5f774dd1a931ce73913e1a4e02b75 and it seems to have fixed things.

It would be better (for Calypso at least) if `@wordpress/core-data` exported an initialization function that we can call at our convenience instead of initializing statically. We can only register the plugins once we know the `userId` (for the persistence plugin) and that's not available yet when modules are statically evaluated. We need to wait at least for the Redux store initialization.

> > Adding the plugins should fix it in theory 🤷‍♂️ Maybe we're adding them too late?
>
> That was my suspicion too, tried moving them before core-data store initialization but no luck so far.

How exactly did you try to do that? Apparently I was more lucky when trying the same thing 🙂

Reply:
> How exactly did you try to do that? Apparently I was more lucky when trying the same thing 🙂

I tried moving the registration call around to get it to execute before `import '@wordpress/core-data';` and added some logging, but completely missed the fact that the above will always be executed first during static initialization. :(

Reply:
Thanks for the fix and detailed explanation Jarda! 🙇

Reply:
Since this is partly a pre-existing problem we've decided to start a follow up issue and handle it in a separate PR.